### ISSUE TITLE
uBlock slows browserbench.org/Speedometer

### ISSUE THREAD

**Issue by Doaxan** [gorhill/uBlock#2552]
With uBlock http://browserbench.org/Speedometer/ benchmark score 34. With/without Adblock Plus score - 52.  It seems to uBlock may slow work of some sites.

- OS/version: Fedora 26 TwentySix/Linux 4.11.0-0.rc6
- Browser/version: Nightly 55.0a1 20170421100201
- uBlock Origin version: 1.12.1

**Comment by gorhill** [comment]
Things to consider:
- It's a best-case scenario for ABP/Firefox.
- It's a worst-case scenario for uBO/Firefox.

Firefox Nightly:
- uBO: 36.7
- uBO w/o cosmetic filtering: 43.1
- ABP: 52.7
- No extension: 53.8

Chrome unstable:
- uBO: 81.8
- uBO w/o cosmetic filtering: 88.1
- ABP: 100.9
- No extension: 102.2

Notice that with Chrome, the difference is not _that_ contrasted despite that specific web page being a best-case scenario for ABP and worst-case scenario for uBO.

Consider: <https://bugzilla.mozilla.org/show_bug.cgi?id=1339572>. The most important thing to keep in mind here, excerpt of one of my comment:

> My current assessment and what benchmarking shows is that uBO's approach it still beneficial overall, for both Chromium and Firefox, despite the edge cases.

If instead of taking one edge case (it's a benchmark which purpose is to keep changing the DOM as fast as possible), you benchmark actual web pages, this is what you get:

<http://www.raymondhill.net/ublock/pageloadspeed.html> with `https://en.wikipedia.org/wiki/List_of_country_calling_codes`:
- Firefox Nightly:
    - uBO: Average (valid): 791.25 ms 
    - ABP: Average (valid): 1488.03 ms
    - No extension: Average (valid): 693.84 ms
- Chrome unstable:
    - uBO: Average (valid): 544.76 ms
    - ABP: Average (valid): 654.99 ms
    - No extension: Average (valid): 566.41 ms

<http://www.raymondhill.net/ublock/pageloadspeed.html> with `http://www.cnn.com/`:
- Firefox Nightly:
    - uBO: Average (valid): 1004.82 ms
    - ABP: Average (valid): 1300.04 ms
- Chrome unstable:
    - uBO: Average (valid): 532.97 ms
    - ABP: Average (valid): 806.61 ms

<http://www.raymondhill.net/ublock/pageloadspeed.html> with `https://www.wired.com/`:
- Firefox Nightly:
    - uBO: Average (valid): 1078.33 ms
    - ABP: Average (valid): 1342.87 ms
- Chrome unstable:
    - uBO: Average (valid): 553.75 ms
    - ABP: Average (valid): 998.97 ms

<http://www.raymondhill.net/ublock/pageloadspeed.html> with `https://www.si.com/`:
- Firefox Nightly:
    - uBO: Average (valid): 1348.90
    - ABP: Average (valid): 1864.59 ms
- Chrome unstable:
    - uBO: Average (valid): 996.70 ms
    - ABP: Average (valid): 1344.74 ms

<http://www.raymondhill.net/ublock/pageloadspeed.html> with `https://alastairreid.github.io/alastairreid.github.io/ARM-v8a-xml-release/`:
- Firefox Nightly:
    - uBO: Average (valid): 84.65 ms
    - ABP: Average (valid): 130.53 ms
- Chrome unstable:
    - uBO: Average (valid): 78.21 ms
    - ABP: Average (valid): 179.72 ms

And all those page load speed gains came with sizable memory saving with uBO vs. ABP.

TL;DR: uBO has edge cases perfomance-wise, but when considering actual real-world cases, it's currently the right approach.

If you still want to reduce overhead of cosmetic filtering on a give site, you can point-and-click to toggle it on a per-site basis (the default can be set to be disabled and enable on a per-site basis). Another approach is to disable entirely generic cosmetic filtering (this is the default in Firefox for Android).

**Comment by gorhill** [comment]
I am going to re-open this following the latest findings, see <http://forums.mozillazine.org/viewtopic.php?p=14758155#p14758155>.

The added line of code mentioned in the above post is to check whether an added node, as reported by the mutation observer, still is part of the DOM tree using a simple `parentNode !== null` test. This simple test helps uBO skip a lot of work during that benchmark, given that the benchmark consist of adding/removing nodes as fast as possible.

I do mention that to test whether a just-added node is still part of the DOM tree when uBO process the outcome of mutation events is likely a pointless test with real world cases: sites just do not add lot of nodes just to remove them all immediately in real world cases.

Scouring the front page of Hacker News, I did actually find a case where uBO ended up dealing with a lot of added nodes which where no longer part of the DOM tree when uBO's handler kicked in, resulting in about ~600 nodes detected as being parentless (hence not in the DOM tree).

Here is the page: <http://cryptoservices.github.io/cryptography/2017/07/21/Sigs.html>.

Turns out that for pages using [MathJax library](https://www.mathjax.org/), the `parentNode === null` condition arise quite a lot for some reasons. Other than this, it's occurring rarely as expected, at most low single-digit number of times whenever it does occur at all.

But even with the MathJax case above, I am undecided as to whether I should add the simple test as the "fix" here. I used quotes for "fix" here because it is still true that speedometer should not be used as an argument that there is a performance issue with uBO.

The added (extra) test here will not yield any measurable improvement for the vast majority of web sites, and does not actual yield measurable improvement even for the one special page above (as reported by Chromium's page load performance tool)[1].

***

[1] Whatever improvement there might be for that one page is completely drowned and made irrelevant by the page's own javascript code -- there is _maybe_ a ~10 ms improvement over a 2,700 ms page load time (I emphasize _maybe_ because whatever is measured might just be below the margin of error of the profiling tool).

**Comment by gorhill** [comment]
Another important point regarding results with ABP: speeDOMeter measures post-`load` event, it does not measures what happens before the `load` event. ABP's overhead of injecting 18K+ CSS rules occurs before the `load` event.
